docs/src/charmil_architecture.md

```diff
@@ -0,0 +1,26 @@
+## Charmil Architecture
+This is how charmil is structred from a high level. 
+![](../images/architecture-diagram.png)
+
+We capture the input from the command line, then process this input. We parse what provider the command is targeting,
+flags and the fields. This data is then passed to the provider, and the respective command is triggered.
+
+### Targeting providers in the cli
+For these examples we will consider a project called `ack` that uses a `print` command to echo a given field that is passed to it.
+
+To run commands that are defined by the `ack` team you use the standard syntax for calling a command.
+```bash
+> ack print Hello
+Hello
+```
+To run commands that are defined by the `charmil` team you use the charmil command and then any sub-command you want.
+```bash
+> ack charmil validate
+All custom commands are valid
+```
+
+### Disabling the charmil provider
+Because `charmil` commands are made for the host and not for the end user you can disable them being available for a release build. 
+```bash
+> go build ack --CHARMIL_COMMANDS=false
```

Comment:
I would avoid that. Lets build separate cli for testing things

Reply:
Updated the docs, rethought how we want to add our commands. In the docs I stated that they are opt-in and have to specifically built for.